### ISSUE TITLE
Remove OpenAPI formats that don't make much sense.

### ISSUE THREAD

**Issue by lavalamp** [kubernetes/kubernetes#97733]
We document (as of https://github.com/kubernetes/kubernetes/pull/85381) that we support the following formats:

```
  // format is an OpenAPI v3 format string. Unknown formats are ignored. The following formats are validated:
  //
  // - bsonobjectid: a bson object ID, i.e. a 24 characters hex string
  // - uri: an URI as parsed by Golang net/url.ParseRequestURI
  // - email: an email address as parsed by Golang net/mail.ParseAddress
  // - hostname: a valid representation for an Internet host name, as defined by RFC 1034, section 3.1 [RFC1034].
  // - ipv4: an IPv4 IP as parsed by Golang net.ParseIP
  // - ipv6: an IPv6 IP as parsed by Golang net.ParseIP
  // - cidr: a CIDR as parsed by Golang net.ParseCIDR
  // - mac: a MAC address as parsed by Golang net.ParseMAC
  // - uuid: an UUID that allows uppercase defined by the regex (?i)^[0-9a-f]{8}-?[0-9a-f]{4}-?[0-9a-f]{4}-?[0-9a-f]{4}-?[0-9a-f]{12}$
  // - uuid3: an UUID3 that allows uppercase defined by the regex (?i)^[0-9a-f]{8}-?[0-9a-f]{4}-?3[0-9a-f]{3}-?[0-9a-f]{4}-?[0-9a-f]{12}$
  // - uuid4: an UUID4 that allows uppercase defined by the regex (?i)^[0-9a-f]{8}-?[0-9a-f]{4}-?4[0-9a-f]{3}-?[89ab][0-9a-f]{3}-?[0-9a-f]{12}$
  // - uuid5: an UUID5 that allows uppercase defined by the regex (?i)^[0-9a-f]{8}-?[0-9a-f]{4}-?5[0-9a-f]{3}-?[89ab][0-9a-f]{3}-?[0-9a-f]{12}$
  // - isbn: an ISBN10 or ISBN13 number string like "0321751043" or "978-0321751041"
  // - isbn10: an ISBN10 number string like "0321751043"
  // - isbn13: an ISBN13 number string like "978-0321751041"
  // - creditcard: a credit card number defined by the regex ^(?:4[0-9]{12}(?:[0-9]{3})?|5[1-5][0-9]{14}|6(?:011|5[0-9][0-9])[0-9]{12}|3[47][0-9]{13}|3(?:0[0-5]|[68][0-9])[0-9]{11}|(?:2131|1800|35\\d{3})\\d{11})$ with any non digit characters mixed in
```

This list is somewhat random and offers some types that I'd strongly recommend folks not use. I'd like to see the following changes, in order of how important it is and how difficult:
* Remove all references to go ("as parsed by..."), this is not helpful to anyone using go, or to anyone who just wants to know what is valid. (only documentation change, hopefully)
* Remove creditcard, password, SSN, maybe email as being unsafe (we have no way of marking fields as PII and omitting them from logs, for example)
* Remove special purpose types, since obviously we can't include all possible special purpose types and we don't let users roll their own, it is very strange to have ISBN in this list.
* personally I think bsonobjectid is also very questionable, both because it's specific to a single database system and because it might be better to represent that as a []byte of length 12 in our API. I'm not sure how common the UUID formats are, they *might* be universal enough to make sense.
* The network types (hostname, uri, ipv4, and ipv6) are universal enough, but is this how we represent them in the rest of our API?

I'm not sure how to fix this without breaking existing users, I'd at least like our documentation to sternly warn against using these features.

/kind bug
/sig api-machinery

**Comment by lavalamp** [comment]
cc @liggitt @sttts @deads2k @DirectXMan12 @apelisse

**Comment by DirectXMan12** [comment]
I think the list of ones that I'd keep, off the top of my head, are

- uri
- ipv4
- ipv6
- cidr
- uuid

In the long term, I think we should add some more k8s-specific ones as well, but that's a separate issue ;-)

Specific comments:

> Remove creditcard, password, SSN, maybe email as being unsafe (we have no way of marking fields as PII and omitting them from logs, for example)

It's pretty much impossible to validate email except by trying to send to it, so email is one of those ones that doesn't make much sense to have IMO, unless you're trying to declare to external tooling "autocomplete from contacts" or something, which might not be as useful to us in k8s, I think.

> The network types (hostname, uri, ipv4, and ipv6) are universal enough, but is this how we represent them in the rest of our API?

Definitely should keep `ipv4` & `ipv6` -- as we start declaring what the validation is for more things in core, we can use these.

`uri` I *think* would similarly be useful for describing the places like webhook where we expect a URI.

`hostname` I'd almost think is worth deprecating/removing in favor of `dns-label` and `dns-subdomain` to match the terminology that Kubernetes uses and actually validates against.

**Comment by lavalamp** [comment]
> Definitely should keep ipv4 & ipv6 

We have these things in our API, in services / endpoints etc. Do they present in OpenAPI like this? (I doubt it)

It's not enough to be useful, if it conflicts with the way we already represent something, that's confusing at best.

**Comment by liggitt** [comment]
In my mind, the point is less how it presents in openapi and more the validation behavior you get for free. Services/endpoints have validation implemented as code. I would definitely expect a CRD wanting an IP field to validate using this format approach rather than add a validating webhook.

**Comment by apelisse** [comment]
Agreed, I'd love it if we had a documented list of accepted format along with:
- Name of the format string,
- Specific validation, possibly as a regex, possibly in the form of code,
- Canonicalization/Comparison function,

Then we would be able to use these for CRDs and built-in.

**Comment by lavalamp** [comment]
I agree that the validation behavior is the important thing. I'd be very surprised if the validation behavior were the same if the openapi presented differently.

**Comment by bunniseng** [comment]
Hey @lavalamp, 

Can I pick this up? Thanks

**Comment by apelisse** [comment]
I think that's part of a much bigger tasks of creating a list of actual useful formats that we use in Kubernetes. That's not easy, but if you're up to it, sure.

**Comment by cici37** [comment]
Hi @apelisse, would you mind verifying if this issue is valid or planned? Thank you
